{% extends 'gecoviz/base.html' %}
{% block content %}
{% load static %}

<div class="container-xxl mx-auto m-1">

    {% include "gecoviz/help/tutorial.html" %}


    <div class="card col-12">
      <div class="card-header card-title mb-0 mt-2 f-bold">
          Searches
      </div>
      <div class="card-body mt-0 m-4">
          <div style="float: left">
              <img src="{% static "gecoviz/img/home.png" %}"
                   alt="Home page"
                   class="mt-2"
                   style="width:95%; height:auto; max-width:65rem"/>
          </div>
          <div class="pt-4 mr-3">
              <div>
                  In the <a href="/" target="_self">home page</a>, you can find two
                  panels, that allow to either:
                  <ul>
                      <li>Browse the KEGG Orthology hierarchy.</li>
                      <li>Enter a custom sequence. After mapping it to eggNOG
                          Orthologous Groups, you can visualize their genomic context.</li>
                  </ul>
              </div>

              <div class="mt-3">
                  On the other hand, explicit searches can be performed by using
                  the <b class="f-bold color-indigo">search bar</b>
                  on the website's navigation bar. Through the selector on
                  the left, choose the category that matches the functional term of
                  interest, which could be entered in the text input on the right.
                  Functional terms may belong to the following categories:
                  <ul>
                      <li>Gene / protein name.</li>
                      <li>Orthologous Groups (KEGG KOs, eggNOG OGs).</li>
                      <li>PFAM protein families.</li>
                  </ul>
                  
              </div>
          </div>
      </div>
    </div>

    <div class="card col-12 mt-3">
      <div class="card-header mb-0 mt-2">
          <div class="card-title f-bold">
              Taxonomy selector
          </div>
          <ul class="nav nav-tabs ml-5" id="ts-tab" role="tablist"
              style="border-bottom:none">
              <li class="nav-item" role="presentation">
                <button class="nav-link active" id="ts-overview-tab" data-toggle="tab"
                    data-target="#ts-overview" type="button" role="tab" 
                    aria-controls="ts-overview" aria-selected="true">
                    Overview
                </button>
              </li>
              <li class="nav-item" role="presentation">
                <button class="nav-link" id="ts-sunburst-tab" data-toggle="tab"
                    data-target="#ts-sunburst" type="button" role="tab" 
                    aria-controls="ts-sunburst" aria-selected="false">
                    Sunburst
                </button>
              </li>
              <li class="nav-item" role="presentation">
                <button class="nav-link" id="ts-basket-tab" data-toggle="tab"
                    data-target="#ts-basket" type="button" role="tab" 
                    aria-controls="ts-basket" aria-selected="false">
                    Genome basket
                </button>
              </li>
          </ul>
      </div>

      <div class="card-body tab-content" id="ts-tab-content">
          <div class="tab-pane fade show active"
              id="ts-overview" role="tabpanel" aria-labelledby="ts-overview-tab">
              <div class="d-flex justify-content-center">
                  <img src="{% static "gecoviz/img/taxselector.png" %}"
                       alt="Genomic context graph outline"
                       class="mt-2"
                       style="width:95%; height:auto; max-width:80rem"/>
              </div>
          </div>
          <div class="tab-pane fade" id="ts-sunburst" 
              role="tabpanel" aria-labelledby="ts-sunburst-tab">
              <div style="float: left">
                  <img src="{% static "gecoviz/img/sunburst.png" %}"
                       alt="Sunburst taxonomy selector"
                       class="mt-2"
                       style="width:95%; height:auto; max-width:65rem"/>
              </div>
              <div class="pt-4 mr-5">
                  An interactive, dynamic, zoomable sunburst acts as the
                  taxonomy selector in <i class="f-bold color-indigo">GeCoViz</i>. 
                  <br><br>
                  Thanks to its hierarchical structure, a sunburst seems an
                  intuitive way to display a phylogeny, in this case, of the
                  genomes in which we find a "hit" (gene associated) to a
                  functional term or orthologous group of interest.
                  <br><br>
                  The sunburst selector is highly interactive and four
                  different items comprise it:
                  <br><br>
                  <b class="color-indigo">Search bar:</b>
                  allows to perform searches on the
                  sunburst. While typing a taxa of interest, a list of
                  suggestions drops down. Choose the one that fits the criteria
                  and the sunburst will dynamically zoom into the selected
                  taxa, highlight it and its submenu will appear (from which
                  you can perform a number of actions, see below).
                  <br><br>
                  <b class="color-indigo">Sunburst:</b> the graph itself is interactive. Hover over
                  a tile and the breadcrumb will update to display the whole
                  taxonomy of the taxa represented by that tile. Click on the
                  tile to toggle the submenu associated with it.
                  <br><br>
                  <b class="color-indigo">Submenu:</b> toggled on a particular taxa. It allows to
                  zoom into such taxa (dynamically modifying the sunburst) as
                  well as to add genomes belonging to it. There are several
                  ways to <b>add genomes to the genome basket</b>. We can
                  choose one genome (at random) that belongs to such taxa, all
                  of them (if it doesn't surpass the limit) or choose the
                  so-called "representative genomes from X taxonomic levels".
                  As an example to the last option, adding 
                  <i>representative genomes from 7 genus</i> means adding
                  one genome (at random) for each of the 7 genus that have
                  matches under the selected taxa.
                  <br><br>
                  <b class="color-indigo">Breadcrumb:</b> located at the bottom of the sunburst. It
                  dynamically portrays the full taxonomy of the taxa that has
                  been selected or hovered on. Moreover, it remains fixed when
                  we zoom in a particular taxa, which is particularly helpful
                  as it allows the user to better orient themselves in the
                  graph.
              </div>
          </div>
          <div class="tab-pane fade" id="ts-basket" 
              role="tabpanel" aria-labelledby="ts-basket-tab">
              <div style="float: left">
                  <img src="{% static "gecoviz/img/basket.png" %}"
                       alt="Genome basket in taxonomy selector"
                       class="mt-2"
                       style="width:95%; height:auto; max-width:50rem"/>
              </div>
              <div class="pt-4 mr-5">
                  The genome basket shows a <b>summary of the selected genomes
                  from different taxa</b>. These genomes can be added from the
                  sunburst's submenu (see sunburst > submenu). A similar
                  submenu can be toggled by clicking on the downwards arrow in
                  any basket entry, allowing the user to modify the number of
                  genomes that will be shown for that taxa. 
                  <br><br>
                  A <b>maximum of 250 genomes</b> can be selected in total.
                  This upper limit is set for user experience, to avoid
                  excessive scrolling when visualizing the genomic context 
                  (a global view is helpful when infering functional
                  conservation) as well as to mantain high browser performance
                  and quick response time to user interaction.
                  <br><br>
                  Once we have finished selecting the desired genomes, we can
                  click on the button on the bottom to visualize their genomic
                  context.
              </div>
          </div>
      </div>
    </div>


    <div class="card col-12 mt-3">
      <div class="card-header mb-0 mt-2">
          <div class="card-title f-bold">
              Genomic context visualizer
          </div>
          <ul class="nav nav-tabs ml-5" id="gcv-tab" role="tablist"
              style="border-bottom:none">
              <li class="nav-item" role="presentation">
                <button class="nav-link active" id="gcv-overview-tab" data-toggle="tab"
                    data-target="#gcv-overview" type="button" role="tab" 
                    aria-controls="gcv-overview" aria-selected="true">
                    Overview
                </button>
              </li>
              <li class="nav-item" role="presentation">
                <button class="nav-link" id="gcv-controlpanel-tab" data-toggle="tab"
                    data-target="#gcv-controlpanel" type="button" role="tab" 
                    aria-controls="gcv-controlpanel" aria-selected="false">
                    Control panel
                </button>
              </li>
              <li class="nav-item" role="presentation">
                <button class="nav-link" id="gcv-graph-tab" data-toggle="tab"
                    data-target="#gcv-graph" type="button" role="tab" 
                    aria-controls="gcv-graph" aria-selected="false">
                    Graph
                </button>
              </li>
              <li class="nav-item" role="presentation">
                <button class="nav-link" id="gcv-legend-tab" data-toggle="tab"
                    data-target="#gcv-legend" type="button" role="tab" 
                    aria-controls="gcv-legend" aria-selected="false">
                    Legend
                </button>
              </li>
          </ul>
      </div>
      <div class="card-body tab-content" id="gcv-tab-content">
          <div class="tab-pane fade show active"
              id="gcv-overview" role="tabpanel" aria-labelledby="gcv-overview-tab">
              <div class="d-flex justify-content-center">
                  <img src="{% static "gecoviz/img/context_vis.png" %}"
                       alt="Genomic context graph outline"
                       class="mt-2"
                       style="width:95%; height:auto; max-width:80rem"/>
              </div>
          </div>
          <div class="tab-pane fade" id="gcv-controlpanel" 
              role="tabpanel" aria-labelledby="gcv-control-tab">
              <div class="d-flex justify-content-center">
                  <img src="{% static "gecoviz/img/control_panel.png" %}"
                       alt="Genomic context control panel"
                       class="mt-2"
                       style="width:95%; height:auto; max-width:80rem"/>
              </div>
              <div class="row row-cols-xxl-3 row-cols-xl-2 row-cols-1 g-2">
                  <div class="col">
                      <div class="card no-box-shadow h-100">
                          <div class="card-header card-title f-bold">
                              1. Tree visualization
                          </div>
                          <div class="card-body p-4 pt-0">
                             The phylogenetic tree on the left of the graph
                             (see overview) can be modified in several ways.
                             Its width can be shrunk (for a more compact
                             visualization) and the branch length and support
                             can be shown or hidden at user's discretion.
                             Lastly, the text displayed on the leaves can be
                             modified. As every leaf node has a taxonomy
                             associated, the taxonomic level that is shown on
                             the tree can be selected.
                          </div>
                      </div>
                  </div>
                  <div class="col">
                      <div class="card no-box-shadow h-100">
                          <div class="card-header card-title f-bold">
                              2. Isolation source
                          </div>
                          <div class="card-body p-4 pt-0">
                              Each genome may have an isolation source
                              associated. We can then filter genomes by their
                              isolation source, by switching on/off the checkboxes shown
                              in section 2 of the figure above. Furthermore,
                              this panel acts as a legend to the isolation
                              source plot.
                              <div class="text-muted mt-3">
                                  Note: genomes with unknown isolation source
                                  can be filtered out by switching off the
                                  "Unknown" checkbox
                              </div>
                          </div>
                      </div>
                  </div>
                  <div class="col">
                      <div class="card no-box-shadow h-100">
                          <div class="card-header card-title f-bold">
                              3. Genomic context
                          </div>
                          <div class="card-body p-4 pt-0">
                              The genomic context visualization (central panel
                              in graph. See a) can be customized in
                              several ways:
                              <ul>
                                  <li>
                                      <b class="f-bold">Gene size:</b> 
                                      gene length and distance can
                                      be scaled to depict genomic dimensions 
                                      (activated by default). We can zoom in
                                      and out by means of the zoom slider.
                                  </li>
                                  <li>
                                      <b class="f-bold">Height collapsing:</b> 
                                      gene height can be collapsed
                                      to fit a greater number of genomes in the
                                      screen. Although less detailed (text on
                                      leaves and genes will not be seen), this
                                      view can be particularly useful if we are
                                      studying a relatively elevated number of
                                      genomes.
                                  </li>
                                  <li>
                                      <b class="f-bold">Number of genes in neighborhood:</b>
                                      the number of genes upstream and downstream
                                      to the anchor/central gene can be
                                      modified independently. A maximum of 10
                                      genes can be visualized on each side.
                                  </li>
                                  <li>
                                      <b class="f-bold">Text on gene:</b>
                                      although the most common
                                      setting is to show their gene name, the
                                      text displayed on genes can be chosen.
                                  </li>
                              </ul>
                          </div>
                      </div>
                  </div>
                  <div class="col">
                      <div class="card no-box-shadow h-100">
                          <div class="card-header card-title f-bold">
                              3. Color by function
                          </div>
                          <div class="card-body p-4 pt-0">
                              Genes can be colored by different functional
                              annotations. This selector allows us to seemingly
                              transition from one to another, obtaining a more
                              comprehensive view of the gene-of-interest
                              neighborhood.
                              Among those functional terms, we can color genes
                              by:
                              <ul>
                                  <li>KEGG pathways</li>
                                  <li>KEGG Orthology</li>
                                  <li>EggNOG Orthology</li>
                                  <li>Pfam domains</li>
                              </ul>
                              Furthermore, hierarchical functional terms such as EggNOG
                              Orthology can be fine-grained using the selector
                              on the right (we can choose at which taxa the
                              orthology cluster has been computed).
                          </div>
                      </div>
                  </div>
                  <div class="col">
                      <div class="card no-box-shadow h-100">
                          <div class="card-header card-title f-bold">
                              4. Conservation threshold
                          </div>
                          <div class="card-body p-4 pt-0">
                              Genes are colored by functional terms. However,
                              to better assess functional conservation, these
                              terms can be filtered by 
                              <b class="f-bold">vertical conservation</b>. 
                              By default, the conservation threshold is set to
                              50% but it can be easily modified by using the
                              slider in the control panel.
                              <div class="text-muted mt-3">
                                  Vertical conservation = genomes with functional term in neighborhood / total genomes
                              </div>
                          </div>
                      </div>
                  </div>
              </div>
          </div>

          <div class="tab-pane fade"
              id="gcv-graph" role="tabpanel" aria-labelledby="gcv-graph-tab">
              <div class="d-flex justify-content-center">
                  <img src="{% static "gecoviz/img/graph.png" %}"
                       alt="Genomic context control panel"
                       class="mt-2"
                       style="width:95%; height:auto; max-width:80rem"/>
              </div>
              <div class="pt-4 mr-5">
                  The dynamic legend for gene color by functional term 
                  is displayed on the right side of the
                  graph (see overview). To hide it, click on the close
                  button "X" on top. Colors can be shuffled by clicking on
                  the corresponding button.
                  <br>
                  <br>
                  Legend entries are functional terms from the category
                  specified on its title, EggNOG Orthology in this example
                  (may be changed from the control panel, section 4). 
                  These entries are ordered by their vertical conservation,
                  and they can be hidden (remove color from genes in
                  the genomic context graph) by clicking on their checkbox
                  between their color and name. 
                  <br>
                  <br>
                  Legend entry names are links to their corresponding
                  website. Vertical conservation and their description (if
                  any) are shown below.
              </div>
          </div>

          <div class="tab-pane fade"
              id="gcv-legend" role="tabpanel" aria-labelledby="gcv-legend-tab">
              <div style="float: left">
                  <img src="{% static "gecoviz/img/legend.png" %}"
                       alt="Genomic context graph outline"
                       class="mt-2"
                       style="width:95%; height:auto; max-width:80rem"/>
              </div>
              <div class="pt-4 mr-5">
                  The dynamic legend for gene color by functional term 
                  is displayed on the right side of the
                  graph (see overview). To hide it, click on the close
                  button "X" on top. Colors can be shuffled by clicking on
                  the corresponding button.
                  <br>
                  <br>
                  Legend entries are functional terms from the category
                  specified on its title, EggNOG Orthology in this example
                  (may be changed from the control panel, section 4). 
                  These entries are ordered by their vertical conservation,
                  and they can be hidden (remove color from genes in
                  the genomic context graph) by clicking on their checkbox
                  between their color and name. 
                  <br>
                  <br>
                  Legend entry names are links to their corresponding
                  website. Vertical conservation and their description (if
                  any) are shown below.
              </div>
          </div>

      </div>
    </div>
</div>


{% endblock %}
